### PR TITLE
api: Use subs for origin in case index is a function

### LINKED ISSUE #1950: Staggering not purged when index contains abs() of a value
If you try to use the absolute value of an index (for example when implementing a free-surface as in the acoustic examples) with a function that is staggered, the staggering is not correctly accounted for during compilation, resulting in a float being used as an index in the generated c code. This causes module compilation to fail.

MFE:
```
import devito as dv

grid = dv.Grid(shape=(11, 11), extent=(10., 10.))
x, y = grid.dimensions
f = dv.TimeFunction(name='f', grid=grid, space_order=2, staggered=x)

eq = dv.Eq(f.forward, f.dx)

lhs, rhs = eq.evaluate.args

funcs = dv.symbolics.retrieve_functions(rhs)
mapper = {}
for f in funcs:
    # Get the x index
    xind = f.indices[1]
    if (xind - x).as_coeff_Mul()[0] < 0:
        mapper.update({f: f.subs({xind: dv.symbolics.INT(abs(xind))})})

eq2 = dv.Eq(lhs, rhs.subs(mapper))

dv.Operator(eq2)(t_M=1)
```

Which returns:
```
/tmp/devito-jitcache-uid1000/1fe24909a093d8c69264ced7adcbb947db766fed.c: In function ‘Kernel’:
/tmp/devito-jitcache-uid1000/1fe24909a093d8c69264ced7adcbb947db766fed.c:47:41: error: array subscript is not an integer
   47 |         f[t1][x + 2][y + 2] = r0*(-f[t0][3.0F/2.0F + (int)(fabs(x - 1.0F/2.0F))][y + 2]) + r0*f[t0][x + 2][y + 2];
      |                                         ^
FAILED compiler invocation:gcc -O3 -g -fPIC -Wall -std=c99 -march=native -Wno-unused-result -Wno-unused-variable -Wno-unused-but-set-variable -ffast-math -shared -fopenmp /tmp/devito-jitcache-uid1000/1fe24909a093d8c69264ced7adcbb947db766fed.c -lm -o /tmp/devito-jitcache-uid1000/1fe24909a093d8c69264ced7adcbb947db766fed.so
```
to the stderr and:
```
Traceback (most recent call last):
  File "staggered_absolute_indices_mfe.py", line 23, in <module>
    dv.Operator(eq2)(t_M=1)
  File "/home/ed/devito/devito/operator/operator.py", line 653, in __call__
    return self.apply(**kwargs)
  File "/home/ed/devito/devito/operator/operator.py", line 724, in apply
    cfunction = self.cfunction
  File "/home/ed/devito/devito/operator/operator.py", line 606, in cfunction
    self._jit_compile()
  File "/home/ed/devito/devito/operator/operator.py", line 591, in _jit_compile
    recompiled, src_file = self._compiler.jit_compile(self._soname,
  File "/home/ed/devito/devito/arch/compiler.py", line 319, in jit_compile
    _, _, _, recompiled = compile_from_string(self, target, code, src_file,
  File "/home/ed/anaconda3/envs/devito/lib/python3.8/site-packages/codepy/jit.py", line 433, in compile_from_string
    toolchain.build_extension(ext_file, source_paths, debug=debug)
  File "/home/ed/anaconda3/envs/devito/lib/python3.8/site-packages/codepy/toolchain.py", line 210, in build_extension
    raise CompileError("module compilation failed")
codepy.CompileError: module compilation failed
```
to the stdout

### PR DESCRIPTION
**Pull request by mloubout**
Fixes #1950

### CODE REVIEW

**Review by FabioLuporini — COMMENTED** [review]
`types` isn't a valid tag anymore; you probably should use `compiler` or `dsl`

File: devito/types/basic.py [hunk 1]
```diff
@@ -1197,8 +1197,8 @@ def indexify(self, indices=None, subs=None):
         subs = [{**{d.spacing: 1, -d.spacing: -1}, **subs} for d in self.dimensions]
 
         # Indices after substitutions
-        indices = [sympy.sympify((a - o).xreplace(s)) for a, o, s in
-                   zip(self.args, self.origin, subs)]
+        indices = [sympy.sympify(a.subs(d, d - o).xreplace(s)) for a, d, o, s in
```
[the remark this answers is not in the document]
Comment (EdCaunt) [reply devito/types/basic.py:1200]: I'm not 100% sure why it was required in the first place unless I'm missing something?

[the remark this answers is not in the document]
Comment (mloubout) [reply devito/types/basic.py:1200]: It's not equivalent no, `a` is the expression for a given dimension of `self` so for `f[x, y+2]`, then `self[x]` is `f[x]` while `a` is just `x`

File: devito/operations/interpolators.py [hunk 2]
```diff
@@ -188,9 +188,16 @@ def _interpolation_indices(self, variables, offset=0, field_offset=0,
                 condition = sympy.And(lb, ub, evaluate=False)
                 mapper[d] = ConditionalDimension(p.name, self.sfunction._sparse_dim,
                                                  condition=condition, indirect=True)
-
+            # Apply mapper to each variable with origin correction before the
+            # dimension get replaced.
```
Comment (EdCaunt) [inline devito/operations/interpolators.py:192]: dimensions

**Review by EdCaunt — COMMENTED** [review]
Would it be worth having a test to double-check that indices are not affected by staggering? Possibly create two simple 1D kernels containing the absolute of an index, one where the function is staggered and one where it is not and check the generated C string is the same.

File: devito/operations/interpolators.py (hunk 2 above)
Comment (mloubout) [reply devito/operations/interpolators.py:192]: fixed

**Comment by mloubout** [comment]
Added test

File: devito/operations/interpolators.py [hunk 3]
```diff
@@ -188,9 +188,16 @@ def _interpolation_indices(self, variables, offset=0, field_offset=0,
                 condition = sympy.And(lb, ub, evaluate=False)
                 mapper[d] = ConditionalDimension(p.name, self.sfunction._sparse_dim,
                                                  condition=condition, indirect=True)
-
+            # Apply mapper to each variable with origin correction before the
+            # dimensions get replaced.
+            vmapper = {}
```
Comment (FabioLuporini) [inline devito/operations/interpolators.py:193]: equivalent to:

```
subs = {v: v.subs({k: c - v.origin.get(k, 0) for k, c in mapper.items()})}
```

?

File: devito/operations/interpolators.py [hunk 4]
```diff
@@ -188,9 +188,16 @@ def _interpolation_indices(self, variables, offset=0, field_offset=0,
                 condition = sympy.And(lb, ub, evaluate=False)
                 mapper[d] = ConditionalDimension(p.name, self.sfunction._sparse_dim,
                                                  condition=condition, indirect=True)
-
+            # Apply mapper to each variable with origin correction before the
```
Comment (FabioLuporini) [inline devito/operations/interpolators.py:191]: Some boring nitpicks here:

* blank line
* use "subs", not "vmapper" (more conventional?)
* no full stop at the end of the comment
* dimensions -> Dimensions

File: devito/operations/interpolators.py (hunk 3 above)
Comment (FabioLuporini) [reply devito/operations/interpolators.py:193]: why does this issue emerge only now?

File: devito/operations/interpolators.py (hunk 4 above)
Comment (mloubout) [reply devito/operations/interpolators.py:192]: done

**Review by FabioLuporini — APPROVED** [review]
OK, thanks

**Pull request merged**